linux/shim/init.go

```diff
@@ -70,11 +73,14 @@ func newInitProcess(context context.Context, path string, r *shimapi.CreateReque
 		IO:            io,
 		NoPivot:       r.NoPivot,
 	}
+	sigchld := make(chan os.Signal, 2048)
+	// main.go calls Notify as well, but it does not hurt
+	signal.Notify(sigchld, syscall.SIGCHLD)
```

Comment:
Thank you, WDYT about this idea:

Move [`cmd/containerd-shim.handleSignals()`](https://github.com/docker/containerd/blob/1d7fa45403128368f21e982d1ad74d40d072e0c4/cmd/containerd-shim/main.go#L114-L130) to `linux/sighandler` as follows:


`linux/sighandler/sighandler.go`:
```go
package sighandler

 // or int maybe
type SignalHandlerID string

// it is valid to call multiple times against same signal
func RegisterSignalHandler(signal os.Signal, handler func()) SignalHandlerID

func UnregisterSignalHandler(id SignalHandlerID)

func HandleSignals(signals chan os.Signal) error 
```


`cmd/containerd-shim/main.go`:
```go
package main

import (
	"github.com/docker/containerd/linux/sighandler"
	...
)

func main(){
	...
	app.Action = func(context *cli.Context) error {
		...
		sighandler.RegisterSignalHandler(syscall.SIGCHLD, func(){
			if err := reaper.Reap(); err != nil {
				logrus.WithError(err).Error("reap exit status")
			}
		})
		sighandler.RegisterSignalHandler(syscall.SIGTERM, func(){
			server.Stop()
		})
		sighandler.RegisterSignalHandler(syscall.SIGINT, func(){
			server.Stop()
		})
		return sighandler.HandleSignals(signals)
	}
}

```

`linux/shim/init.go`
```go
import (
	"github.com/docker/containerd/linux/sighandler"
	...
)

func receiveConsoleSocketMaster(socket *runc.ConsoleSocket, deadline time.Duration) (console.Console, error) {
	var cons console.Console
	var err error
	sigchldChan := make(chan os.Signal)
	sigchldHandler := sighandler.RegisterSignalHandler(func() {
		sigchldChan <- os.Signal
	})
	defer sighandler.UnregisterSignalHandler(sigchldHandler)
	done := make(chan struct{}, 1)
	go func() {
		cons, err = socket.ReceiveMaster()
		done <- struct{}{}
	}()
	select {
	case s := <-sigchldChan:
		err = errors.Errorf("got signal while receiving pty master: %v", s)
	case <-time.After(deadline):
		err = errors.Errorf("exceeded deadline %v while receiving pty master", deadline)
	case <-done:
	}
	return cons, err
}

```